crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -393,27 +393,27 @@ where
                 self.visit_expr(&node.test);
                 let pre_if = self.flow_snapshot();
                 self.visit_body(&node.body);
-                let mut last_clause_is_else = false;
-                let mut post_clauses: Vec<FlowSnapshot> = vec![self.flow_snapshot()];
+                let mut has_else = false;
+                let mut post_clauses: Vec<FlowSnapshot> = vec![];
                 for clause in &node.elif_else_clauses {
-                    // we can only take an elif/else clause if none of the previous ones were taken
+                    // snapshot after every block except the last; the last one will just become
+                    // the state that we merge the other snapshots into
+                    post_clauses.push(self.flow_snapshot());
+                    // we can only take an elif/else branch if none of the previous ones were
+                    // taken, so the block entry state is always `pre_if`
                     self.flow_restore(pre_if.clone());
                     self.visit_elif_else_clause(clause);
-                    post_clauses.push(self.flow_snapshot());
                     if clause.test.is_none() {
-                        last_clause_is_else = true;
+                        has_else = true;
                     }
                 }
-                let mut post_clause_iter = post_clauses.into_iter();
-                if last_clause_is_else {
-                    // if the last clause was an else, the pre_if state can't directly reach the
-                    // post-state; we must enter one of the clauses.
-                    self.flow_restore(post_clause_iter.next().unwrap());
-                } else {
-                    self.flow_restore(pre_if);
+                for post_clause_state in post_clauses {
+                    self.flow_merge(&post_clause_state);
                 }
-                for post_clause_state in post_clause_iter {
-                    self.flow_merge(post_clause_state);
+                if !has_else {
```

Comment:
Nit: Or, to avoid a write in every loop and make it a constant access instead

```suggestion
                let has_else = node.elif_else_clauses.last().is_some_and(|clause| clause.test.is_none());
                if !has_else {
```